### PR TITLE
Enumerable

### PR DESCRIPTION
**Pull request by mikenath223**

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Try :+1: 
You have done a splendid job implementing the requirements for this project. I tested your methods and they pass most edge cases except a few which we need to fix so that we can merge.
Check my comments for suggestions on how to fix the failing edge cases.

### Martin Afani

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    return to_enum unless block_given?
+
+    entry = is_a(Range) ? to_a : self
+    i = 0
+    while i < size
+      yield(entry[i], i)
+      i += 1
+    end
+    entry
+  end
```
Comment (whiz25) [inline main.rb]: - [x] calls the given block once for each element in self

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      true
+    end
+    if arg.nil?
+      my_each { |elem| return false unless elem }
+      return true
+    end
+
+    my_each { |elem| return false unless check_validity(elem, arg) }
+    true
+  end
```
Comment (whiz25) [inline main.rb]: - [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+  def my_any?(arg = nil, &proc)
+    if block_given?
+      my_each { |elem| return true if proc.nil ? proc.call(elem) : yield(elem) }
+    else
+      my_each { |elem| return true if arg.nil ? elem : check_validity(elem, arg) }
+    end
+    false
+  end
```
Comment (whiz25) [inline main.rb]: - [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+  def my_none?(pattern = nil, &proc)
+    !my_any?(pattern, &proc)
+  end
```
Comment (whiz25) [inline main.rb:59]: - [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x]  when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

**Comment by mikenath223** [comment]
Thanks for this. I  have implemented corrections as instructed.

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Awesome @mikenath223 :+1: 
You implemented some of the changes as I suggested, that is good. However, some edge cases are still missing. Please see my comments and try your best to implement them.

### Martin Afani

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      my_each { |elem| return false unless yield(elem) }
+      return true
+    end
+    if arg.nil?
+      my_each { |elem| return false unless elem }
+      return true
+    end
+    my_each { |elem| return false unless check_validity(elem, arg) }
+    true
+  end
```
Comment (whiz25) [inline main.rb]: **The following are still missing**
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x]  when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

**Comment by mikenath223** [comment]
Oh thanks. Please help me outline the aspects that still remains. I would appreciate that.

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+  def my_any?(arg = nil, &proc)
+    if block_given?
+      my_each { |elem| return true if proc.nil? ? proc.call(elem) : yield(elem) }
+    else
+      my_each { |elem| return true if arg.nil? ? elem : check_validity(elem, arg) }
+    end
+    false
+  end
```
Comment (whiz25) [inline main.rb:55]: **These too**
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+  def my_none?(pattern = nil, &proc)
+    !my_any?(pattern, &proc)
+  end
```
Comment (whiz25) [inline main.rb:59]: **Finally**
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :+1: :tada: 
Great job implementing these changes. Your project is approved :smile: 
You can merge it now
Good luck with the next one.
Happy conding :computer:

**Comment by mikenath223** [comment]
Thanks for this. Lovely

**Pull request merged**